sympy/physics/quantum/lapply.py

```diff
@@ -0,0 +1,1434 @@
+# -----------------------------------------------------------------------------
+# The purpose of caching is to avoid frequent recomputation if a Pow is
+# rolled off to Pow(base,exp-1)*base -> Pow(base,exp-2)*base*base etc.
+#
+# Principles of Caching for base and exponents of Powers:
+# - Only cache exprs that will probably be used multiple times. Do not litter the
+#   the cache with one-time results. So bases of Powers are cached. Exponents are
+#   cached only in case of option["apply_exp"] and non-simple exponents.
+# - Exprs being cached must have been returned in ncl from lapply_mul2_c_nc, so
+#   all its factors have been processed fully.
+# - The cache returns three types of information on an expr (besides None):
+#   i.   "elem": expr is elementary  ii. "expr": expr is fully processed
+#   iii. (rolloff, cl, encl) expr is base of a power.
+#   which roughly correspond to output of c_nc_ncef, lapply_mul2, lapply_mul2_c_nc.
+# - Base and exponents of Pows are cached separately. If a power is rolled off
+#   the exponents counted down are removed from cache to avoid litter. This may
+#   lead to re-computation of exponents in situations as Pow(A,exp)*B+Pow(A,exp)*C
+#   which is unavoidable as the cache only has one instance.
+
+class RO(Enum): # the roll-off type of a base:
+    UNKN = 0    # Roll-off type is unknown
+    NORO = 1    # base*base doesn't interact, so don't roll off base**exp
+    ROLL = 2    # base*base interacts, so roll off base**exp off
+
+# Pow_with_Cache returns Pow(base, exp), but also caches the rolloff
+# decision assigned to base (rolloff) and the splitting of base
+# into cl and elementary ncl factors (as derived from lapply_mul2_c_nc).
+def Pow_with_Cache(base:Expr, exp:Union[Expr, None], rolloff:RO,
+                   cl:List, encl:List, **options) -> Union[Expr, Tuple]:
+    """Caches factorization data for base. If exp is None, returns the
+    cached tuple (rolloff, cl, encl) else Pow(base, exp)."""
+
+    if exp is not None:  # exp==None is to register a base only
+        if   exp is S.Zero: p = S.One
+        elif exp is S.One : p = base
+        else: p = Pow(base, exp, evaluate=False)
+
+    if base.is_commutative: # in this case overwrite any other rolloff
+        # note that factors in encl may be nc but give a c product!
+        rolloff, cl, encl = RO.ROLL, [base], [] # let Mul handle base.
+
+    to_cache = (rolloff, cl.copy(), encl.copy())
+    cache = options["cache_for_pow"]
+    # The cache is indexed on base, as only attributes of base are stored.
+    if (cached := cache.setdefault(base, to_cache)) == to_cache:
+        return cached if exp is None else p
+    if type(cached) is not tuple: # cache[base] is "elem" or "expr"
+        cached = to_cache         # but cache for base has priority
+
+    # Rolloff may not be downgraded to UNKN
+    if rolloff == RO.UNKN: rolloff = cached[0]
+
+    cache[base] = (rolloff, to_cache[1], to_cache[2])
+    return cached if exp is None else p

```

Comment:
We get here if `cached == cached[base]` so I guess this is modifying the cache entry or something. That seems like quite complicated control flow.

Reply:
Good point. I have shifted the caching into a separate function cache_for_base().  The structure should be much clearer now. 

The cache is incremental in the sense that if additional information about a base becomes available that cache entry is updated.